base/bundle.cpp

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include "base/bundle.hpp"
+
+#include <functional>
+#include <list>
+
+#include "base/map_util.hpp"
+#include "base/status.hpp"
+
+namespace principia {
+namespace base {
+
+thread_local std::function<bool()> AbortRequested = [] { return false; };
+
+Bundle::TaskHandle::TaskHandle(std::shared_ptr<Task> task)
+    : task_(std::move(task)) {}
+
+Bundle::Bundle(int threads)
+    : terminate_on_empty_(false),
+      max_workers_(threads) {
+  workers_.reserve(max_workers_);
+}
+
+Status Bundle::Join() {
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    CHECK(!terminate_on_empty_);
+    terminate_on_empty_ = true;
+  }
+  tasks_not_empty_or_terminate_.notify_all();
+  for (auto& thread : workers_) {
+    thread.join();
+  }
+  std::unique_lock<std::mutex> lock(lock_);
+  return status_;
+}
+
+Bundle::TaskHandle Bundle::Add(std::function<Status()> task_body) {
+  if (workers_.size() < max_workers_) {
+    workers_.emplace_back(&Bundle::Toil, this);
+  }
+  // Private constructor, cannot |make_shared|.
+  TaskHandle result(std::make_shared<Task>(this, std::move(task_body)));
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    CHECK(!terminate_on_empty_);
+    tasks_.emplace_back(result.task_);
+    auto end = tasks_.end();
+    result.task_->set_queue_it(--end);
+  }
+  tasks_not_empty_or_terminate_.notify_one();
+  return result;
+}
+
+void Bundle::Abort(TaskHandle const& handle) {
+  not_null<Task*> task = handle.task_.get();
+  task->Abort();
+}
+
+Status Bundle::JoinTask(TaskHandle const handle) {
+  not_null<Task*> task = handle.task_.get();
+  task->wait_for_termination();
+  return task->status();
+}
+
+void Bundle::Toil() {
+  std::shared_ptr<Task> current_task;
+  for (;;) {
+    {
+      std::unique_lock<std::mutex> lock(lock_);
+      if (terminate_on_empty_ && tasks_.empty()) {
```

Comment:
I am going to claim that you don't need this statement, or rather that you can merge it with the one at line 78.  You can always wait on the condition variable, and if you must exit the loop it will immediately be true.